gtsfm/evaluation/visualize_benchmark_comparison.py

```diff
@@ -0,0 +1,227 @@
+"""
+Use Plotly to create a dashboard that compares the metrics across all the benchmarks from the CI.
+
+The dashboard is a heatmap representing a 2d table, with text annotations added to it.
+
+Authors: John Lambert
+"""
+
+import argparse
+from collections import defaultdict
+from pathlib import Path
+from typing import List
+
+import numpy as np
+import plotly.graph_objects as go
+from matplotlib import colors
+from matplotlib.colors import LinearSegmentedColormap
+from plotly.graph_objs.layout import Annotation, Font, Margin, XAxis, YAxis
+
+import gtsfm.evaluation.merge_reports as report_utils
+import gtsfm.evaluation.metrics_report as metrics_report
+import gtsfm.utils.metrics as metrics_utils
+
+
+HEATMAP_WIDTH = 1500
+HEATMAP_HEIGHT = 900
+NUM_COLORS_COLORMAP = 100
+MAX_NUM_CHARS_ARTIFACT_FNAME = 35
+
+MIN_RENDERABLE_PERCENT_CHANGE = -20
+MAX_RENDERABLE_PERCENT_CHANGE = 20
+
+DASHBOARD_HTML_SAVE_FPATH = Path(__file__).parent.parent.parent / "visual_comparison_dashboard.html"
+
+ZIP_FNAMES = [
+    "deep_front_end-2011205_rc3-20-png-wget-astronet-1024-true.zip",
+    "deep_front_end-door-12-12-JPG-test_data-olsson-loader-1296-true.zip",
+    "deep_front_end-notre-dame-20-20-jpg-gdrive-colmap-loader-760-false.zip",
+    "deep_front_end-skydio-8-8-jpg-gdrive-colmap-loader-760-true.zip",
+    "deep_front_end-skydio-32-32-jpg-gdrive-colmap-loader-760-true.zip",
+    "sift_front_end-2011205_rc3-65-png-wget-astronet-1024-true.zip",
+    "sift_front_end-door-12-12-JPG-test_data-olsson-loader-1296-true.zip",
+    "sift_front_end-palace-fine-arts-281-25-jpg-wget-olsson-loader-320-true.zip",
+    "sift_front_end-skydio-8-8-jpg-gdrive-colmap-loader-760-true.zip",
+    "sift_front_end-skydio-32-32-jpg-gdrive-colmap-loader-760-true.zip",
+]
+
+TABLE_NAMES = [
+    "Verifier Summary",
+    "Inlier Support Processor Summary",
+    "Rotation Cycle Consistency Metrics",
+    "Cycle Consistent Frontend Summary",
+    "Averaging Metrics",
+    "Data Association Metrics",
+    "Bundle Adjustment Metrics",
+]
+
+RED_HEX = "#df0101"
+PALE_YELLOW_HEX = "#f5f6ce"
+GREEN_HEX = "#31b404"
+
+
+def colorscale_from_list(colorlist: List[str]) -> List[str]:
+    """Create hex colorscale to interpolate between requested colors.
+
+    Args:
+        colorlist: requested colors.
+
+    Returns:
+        colorscale: list of length (NUM_COLORS_COLORMAP+1) representing a list of colors.
+    """
+    cmap = LinearSegmentedColormap.from_list(name="dummy_name", colors=colorlist)
+    colorscale = [colors.rgb2hex(cmap(k * 1 / NUM_COLORS_COLORMAP)) for k in range(NUM_COLORS_COLORMAP + 1)]
+    return colorscale
+
+
+def plot_colored_table(X: List[str], Y: List[str], Z: np.ndarray) -> str:
+    """Create an annotated heatmap.
+
+    Args:
+        X: labels for each column (column names).
+        Y: labels for each row (row names).
+        Z: 2d matrix, representing percentage changes from value for a metric on the master branch.
```

Comment:
These variable names don't represent the information they hold very well. Maybe `row_labels`, `col_tables`, `data` or something would be clearer.

Reply:
good point, updated now.